artemis-server/src/main/java/org/apache/activemq/artemis/core/paging/cursor/impl/LivePageCacheImpl.java

```diff
@@ -48,54 +82,228 @@ public long getPageId() {
       }
-      this.messages.add(message);
+      while (true) {
+         final long pIndex = producerIndex;
+         if (pIndex != RESIZING) {
+            if (pIndex == Integer.MAX_VALUE) {
+               throw new IllegalStateException("can't add more then " + Integer.MAX_VALUE + " messages");
+            }
+            //load acquire the current producer buffer
+            final AtomicChunk<PagedMessage> producerBuffer = this.producerBuffer;
+            final int pOffset = (int) (pIndex & chunkMask);
+            //only the first message to a chunk can attempt to resize
+            if (pOffset == 0) {
+               if (appendChunkAndMessage(producerBuffer, pIndex, message)) {
+                  return;
+               }
+            } else if (PRODUCER_INDEX_UPDATER.compareAndSet(this, pIndex, pIndex + 1)) {
+               //this.producerBuffer is the correct buffer to append a message: it is guarded by the producerIndex logic
+               //NOTE: producerIndex is being updated before setting a new value
+               producerBuffer.lazySet(pOffset, message);
+               return;
+            }
+         }
+         Thread.yield();
+      }
+   }
+
+   private boolean appendChunkAndMessage(AtomicChunk<PagedMessage> producerBuffer, long pIndex, PagedMessage message) {
+      if (!PRODUCER_INDEX_UPDATER.compareAndSet(this, pIndex, RESIZING)) {
+         return false;
+      }
+      final AtomicChunk<PagedMessage> newChunk;
+      try {
+         final int index = (int) (pIndex >> chunkSizeLog2);
+         newChunk = new AtomicChunk<>(index, producerBuffer, chunkSize);
+      } catch (OutOfMemoryError oom) {
+         //unblock producerIndex without updating it
+         PRODUCER_INDEX_UPDATER.lazySet(this, pIndex);
+         throw oom;
+      }
+      //adding the message to it
+      newChunk.lazySet(0, message);
+      //linking it to the old one, if any
+      if (producerBuffer != null) {
+         //a plain store is enough, given that producerIndex prevents any reader/writer to access it
+         producerBuffer.next = newChunk;
+      } else {
+         //it's first one
+         this.consumerBuffer = newChunk;
+      }
+      //making it the current produced one
+      this.producerBuffer = newChunk;
+      //store release any previous write and "unblock" anyone waiting resizing to finish
+      PRODUCER_INDEX_UPDATER.lazySet(this, pIndex + 1);
+      return true;
    }
 
    @Override
-   public synchronized void close() {
+   public void close() {
       logger.tracef("Closing %s", this);
       this.isLive = false;
    }
 
+   private static PagedMessage[] EMPTY_MSG = null;
+
+   private static PagedMessage[] noMessages() {
+      //it is a benign race: no need strong initializations here
+      PagedMessage[] empty = EMPTY_MSG;
+      if (empty != null) {
+         return empty;
+      } else {
+         empty = new PagedMessage[0];
```

Comment:
Why not simply make noMessages return a static empty array, e.g. make PagedMessage[0] - similar in nature to https://android.googlesource.com/platform/libcore/+/jb-mr2-release/luni/src/main/java/libcore/util/EmptyArray.java, and then saves instantiating every time.  

Also avoids a race condition, this current code, seems to have with lazy initialization, e.g. currently if two threads concurrently call noMessages its possible i get two PagedMessage[0] objects created.

Reply:
yep I have done it on purpose: is a benign race, because there is just a very low chance a new empty would be allocated several times, but I agree :+1: 
It will makes the code much easier